### ISSUE TITLE
Significant Performance Discrepancy Between GPU and CPU

### ISSUE THREAD

**Issue by DAigory** [google-deepmind/mctx#96]
Hi 
I am currently testing the examples/policy_improvement_demo.py script and have noticed a significant performance discrepancy between GPU and CPU. Specifically, the runtime on GPU is approximately 10 times slower than on CPU.

Here is a snippet of the code I'm using: it is jast wrap of main policy_improvement_demo.py
jitted_run_demo = jax.jit(_run_demo, backend="cpu") #vs gpu about 10times and cpu faster
runner_state, traj_batch = jax.lax.scan(
    jitted_run_demo, (rng_key, 0.0), None, 20000  # config["TIME_STEP_COLLECT"]
)

**Comment by fidlej** [comment]
Some hints:
The demo does not use large matrices. It is possible that the demo runs faster on CPU.

**Comment by DAigory** [comment]
I have a question regarding the performance and implementation details of MCTS (Monte Carlo Tree Search) and JAX:

Is it correct to conclude that the MCTS tree is always faster when executed on a CPU? Additionally, does the choice of JAX for implementation prevent hybrid computations, such that recurrent_fn and rootn_fn cannot be executed on a GPU, and params never leave the GPU memory?

Is it true that JAX does not support hybrid memory and function calls in the context of a task where I want root and recurrent_fn to be executed on the GPU, and each call returns only part of the output (reward/value logits) to the CPU for tree computation, while next_embedded remains in GPU memory along with env, the buffer, and trainState? If I am correct and JAX does not support this, might another framework be more suitable, and thus the choice of JAX for MCTS tasks might be incorrect?

Looking forward to your insights on this matter.

Thanks!

**Comment by fidlej** [comment]
JAX supports execution on multiple devices. But it would be probably faster to keep all execution on the same device. The transfer of the data from one device to another device adds a latency. You can measure the speed of your network on your GPU.

**Comment by DAigory** [comment]
Why do you believe that transferring data from GPU to CPU can slow down the process if all data should always remain on the GPU, and only value, reward, and logits need to be returned to the CPU at each MCTS step? In this case, even embedding might not be necessary, and just references to GPU memory would suffice.

I assumed that parameters, the environment, the model, and other resources could remain on the GPU, but I have not been able to find such a possibility in JAX for clear control over this process. I would appreciate an explanation on this matter.

Thank you!

**Comment by DAigory** [comment]
As far as I understand, any functions executed under jax.scan, jax.fori_loop, etc., expect all arguments to be on the same device, regardless of where the data was allocated at the time of compilation with jax.jit. They do not allow combining their location. Since the inference functions I want to run on the GPU are executed within the mctx function, which uses JAX loops, I am forced to transfer all arguments to the same device where mctx itself is compiled. This means I cannot hybridly execute mctx on the CPU and the inference model on the GPU within the JAX framework.

As I am not an expert in JAX, could you please confirm if this understanding is correct?

**Comment by fidlej** [comment]
Thanks for mentioning the jax.scan and jax.fori_loop properties. Really, I think you will want to run everything on the same device, if possible. In any framework, data transfer between the CPU and GPU will add latency.
You would see latency when profiling: https://jax.readthedocs.io/en/latest/profiling.html
